Tzkt.Api/Websocket/Processors/CyclesProcessor.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.Logging;
+
+using Tzkt.Api.Repositories;
+using Tzkt.Api.Services.Cache;
+using Tzkt.Api.Models;
+
+namespace Tzkt.Api.Websocket.Processors
+{
+    public class CyclesProcessor<T> : IHubProcessor where T : Hub
+    {
+        #region static
+        const string CycleGroup = "cycles";
+        const string CycleChannel = "cycles";
+        static readonly SemaphoreSlim Sema = new(1, 1);
+        static readonly Dictionary<int, HashSet<string>> DelaySubs = new();
+        static readonly Dictionary<string, Symbols> QuoteSubs = new();
+        #endregion
+
+        readonly StateCache StateCache;
+        readonly CyclesRepository CyclesRepo;
+        readonly IHubContext<T> Context;
+        readonly ILogger Logger;
+
+        public CyclesProcessor(StateCache cache, CyclesRepository repo, IHubContext<T> hubContext, ILogger<CyclesProcessor<T>> logger)
+        {
+            StateCache = cache;
+            CyclesRepo = repo;
+            Context = hubContext;
+            Logger = logger;
+        }
+
+        public async Task OnStateChanged()
+        {
+            var sendings = new List<Task>(2);
+            try
+            {
+                await Sema.WaitAsync();
+
+                var cycleData = await CyclesRepo.Get(StateCache.Current.Cycle, Symbols.None);
```

Comment:
There is no need to do DB request every block, because cycles are update once per 3 days. I'd suggest something like this:
```cs
 public class CyclesProcessor<T> : IHubProcessor where T : Hub
 {
     ...
     Cycle CurrentCycle = null;
     ...
     
     public async Task OnStateChanged()
     {
         ...
         if (CurrentCycle == null || StateCache.Current.Level < CurrentCycle.FirstLevel || StateCache.Current.Level > CurrentCycle.LastLevel)
             CurrentCycle = await CyclesRepo.Get(StateCache.Current.Cycle, Symbols.None);
```

Reply:
Added. 🙂